mlflow/types/schema.py

```diff
@@ -116,84 +118,189 @@ def __repr__(self) -> str:
+                    tuple, shape.__class__
+                )
+            )
+        self._type = type
+        self._shape = tuple(shape)
+
+    @property
+    def type(self) -> np.dtype:
+        """
+        A unique character code for each of the 21 different numpy built-in types.
+        See https://numpy.org/devdocs/reference/generated/numpy.dtype.html#numpy.dtype for details.
+        """
+        return self._type
+
+    @property
+    def name(self) -> Optional[str]:
+        """The tensor name or None if the tensor is unnamed."""
+        return self._name
+
+    @property
+    def shape(self) -> tuple:
+        """The tensor shape"""
+        return self._shape
+
+    def to_dict(self) -> Dict[str, Any]:
+        if self.name is None:
+            return {"type": self.type.name, "shape": self.shape}
+        else:
+            return {"name": self.name, "type": self.type.name, "shape": self.shape}
+
+    @classmethod
+    def from_json_dict(cls, **kwargs):
+        """
+        Deserialize from a json loaded dictionary.
+        The dictionary is expected to contain `type` and `shape` keys.
+        """
+        tensor_type = np.dtype(kwargs["type"])
+        tensor_shape = tuple(kwargs["shape"])
+        return cls(tensor_type, tensor_shape, kwargs["name"] if "name" in kwargs else None)
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, TensorSpec):
+            names_eq = (self.name is None and other.name is None) or self.name == other.name
+            return names_eq and self.type == other.type and self.shape == other.shape
+        return False
+
+    def __repr__(self) -> str:
+        if self.name is None:
+            return "({type}, {shape})".format(type=repr(self.type.name), shape=repr(self.shape))
+        else:
+            return "({name}, {type}, {shape})".format(
+                name=repr(self.name), type=repr(self.type.name), shape=repr(self.shape)
+            )
+
+
 class Schema(object):
     """
-    Specification of types and column names in a dataset.
+    Specification of a dataset.
+
+    Schema is represented as a list of :py:class:`ColSpec` or :py:class:`TensorSpec`. A combination
+    of `ColSpec` and `TensorSpec` is not allowed.
 
-    Schema is represented as a list of :py:class:`ColSpec`. The columns in a schema can be named,
-    with unique non empty name for every column, or unnamed with implicit integer index defined by
-    their list indices. Combination of named and unnamed columns is not allowed.
+    The dataset represented by a schema can be named, with unique non empty names for every input.
+    In the case of :py:class:`ColSpec`, the dataset columns can be unnamed with implicit integer
+    index defined by their list indices.
+    Combination of named and unnamed data inputs are not allowed.
     """
 
-    def __init__(self, cols: List[ColSpec]):
+    def __init__(self, dataset: List[Union[ColSpec, TensorSpec]]):
```

Comment:
nit: `dataset` -> `inputs` ? I think that is more clear since you access it as `inputs` after.